lib/galaxy/managers/interactivetool.py

```diff
@@ -18,125 +19,129 @@
 )
 from galaxy.model.base import transaction
 from galaxy.security.idencoding import IdAsLowercaseAlphanumEncodingHelper
-from galaxy.util.filelock import FileLock
```

Comment:
I feel like I may be missing something regarding the `FileLock`. What role was it actually playing? Isn't SQLite [already taking care of locks](https://www.sqlite.org/lockingv3.html)?

> The pager module effectively controls access for separate threads, or separate processes, or both. Throughout this document whenever the word "process" is written you may substitute the word "thread" without changing the truth of the statement.

Reply:
Galaxy has a long history of concurrency issues with respect to SQLite. I cannot promise this FileLock was needed but experience taught me it was better to be safe than sorry here.

Reply:
Some information that may help us understand whether the lock is needed in this case.

- [SQLALchemy documentation](https://docs.sqlalchemy.org/en/20/dialects/sqlite.html#database-locking-behavior-concurrency)
> However, no matter what locking modes are used, SQLite will still always lock the database file once a transaction is started and DML (e.g. INSERT, UPDATE, DELETE) has at least been emitted, and this will block other transactions at least at the point that they also attempt to emit DML. By default, the length of time on this block is very short before it times out with an error.
>
> This behavior becomes more critical when used in conjunction with the SQLAlchemy ORM. SQLAlchemy’s [Session](https://docs.sqlalchemy.org/en/20/orm/session_api.html#sqlalchemy.orm.Session) object by default runs within a transaction, and with its autoflush model, may emit DML preceding any SELECT statement. This may lead to a SQLite database that locks more quickly than is expected. The locking mode of SQLite and the pysqlite driver can be manipulated to some degree, however it should be noted that achieving a high degree of write-concurrency with SQLite is a losing battle.

- [SQLite documentation](https://www.sqlite.org/whentouse.html)
> # 2. Situations Where A Client/Server RDBMS May Work Better
>
> - [...]
>
> - **High-volume Websites**
>
>   SQLite will normally work fine as the database backend to a website. But if the website is write-intensive or is so busy that it requires multiple servers, then consider using an enterprise-class client/server database engine instead of SQLite.
>
>
> - [...]
>
> - **High Concurrency**
>   SQLite supports an unlimited number of simultaneous readers, but it will only allow one writer at any instant in time. For many situations, this is not a problem. Writers queue up. Each application does its database work quickly and moves on, and no lock lasts for more than a few dozen milliseconds. But there are some applications that require more concurrency, and those applications may need to seek a different solution. 

From both excerpts I understand that it is likely that the long history of concurrency issues that Galaxy has with SQLite probably has to do with using the SQLAlchemy ORM under high load (maybe you can either support/refute this?).

As far as I know, the interactive tools mapping file is written only by `InteractiveToolSqlite` (or `InteractiveToolPropagatorSQLAlchemy` in this refactor) and read only by Galaxy (the same class) and the proxy. I understand that issues would appear if interactive tools are in extremely high demand (Galaxy constantly writes keys and tokens to the file). However, in that case, wouldn't it just make better sense, as the SQLite docs claim, that the Galaxy admin uses PostgreSQL rather than SQLite?

Maybe if @jdavcs agrees to [go forward with a new config property for mappings based on arbitrary SQLAlchemy database URLs](https://github.com/galaxyproject/galaxy/pull/18481#issuecomment-2222244009), it may even make sense to discourage the use of the original property `interactivetools_map` in the Galaxy docs to favour the use of Postgres? After all, [SQLite use is already discouraged](https://docs.galaxyproject.org/en/release_24.1/admin/production.html#switching-to-postgresql-database-server).

In that case though, I think we should also have a look together at the [proxy-side implementation](https://github.com/galaxyproject/gx-it-proxy/pull/21/files#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5R24-R53), because it may also have undesired drawbacks.

Reply:
> From both excerpts I understand that it is likely that the long history of concurrency issues that Galaxy has with SQLite probably has to do with using the SQLAlchemy ORM under high load (maybe you can either support/refute this?).

I think the core of the issue here is just SQLite: "SQLite will still always lock the database file once a transaction is started" - it's just suboptimal for write-intensive scenarios. Using the ORM makes it worse: the autoflush model mentioned in the docs ensures that the state of objects loaded from the db is kept in sync with the data in the db by issuing DML statements before retrieving new data (when needed) - so the db file gets locked much more frequently if we rely on the Session. However, we don't use the ORM here.

If we go with [this proposed approached](https://github.com/galaxyproject/galaxy/pull/18481#issuecomment-2222244009), we can still keep SQLite as an option (it may be still a good enough option for smaller instances or dev purposes, etc.; after all we support SQLite as an option for the main database too). We may consider adding a note to [Galaxy's documentation](https://docs.galaxyproject.org/en/master/admin/special_topics/interactivetools.html#interactivetools) and updating the [IT tutorial](https://training.galaxyproject.org/training-material/topics/admin/tutorials/interactive-tools/tutorial.html) accordingly.

Reply:
I added a note (dddf5ba) to the documentation and opened galaxyproject/training-material#5178 and galaxyproject/training-material#5179